### ISSUE TITLE
Support for Hoppers with Movable Hurdles

### ISSUE THREAD

**Issue by vlaicu** [thomas-maeder/popeye#335]
Please provide support for Bul and Dob pieces.
The definitions and examples can be found in the uploaded file.
Thank you very much!
[ANDA-Fairy-Planet-No-1-2020.pdf](https://github.com/thomas-maeder/popeye/files/8499600/ANDA-Fairy-Planet-No-1-2020.pdf)

**Comment by thomas-maeder** [comment]
@vlaicu I have programmed Bul hoppers. Please see the output for the example problems:

- https://github.com/thomas-maeder/popeye/blob/382-feature-request-bul-pieces/REGRESSIONTESTS/bul.tst
- https://github.com/thomas-maeder/popeye/blob/382-feature-request-bul-pieces/REGRESSIONTESTS/bul.ref

**Comment by vlaicu** [comment]
Thank you very much, @thomas-maeder!
Cooks found in Crisan / Luce: the promotions in LI / G are not legal. Only promotions in bLI / bG are legal.

**Comment by thomas-maeder** [comment]
`the promotions in LI / G are not legal.`
Why wouldn't they be legal?

**Comment by thomas-maeder** [comment]
IMHO, this is absurd - not going to happen in Popeye.
Condition PromOnly can be used to restrict the types of promotees.

**Comment by vlaicu** [comment]
I understand your opinion. 
The example with HurdleColourChanging specification is the easiest to give because it applies only to hoppers.
In the H#1 given above, Popeye doesn't consider 1... g:h8=G# as valid. Only 1... g:h8=hccG# is given as a solution.
Does Popeye consider a fairy piece only by looking at the type?
Or does Popeye consider a fairy piece by considering the pair (specification, type)?

**Comment by thomas-maeder** [comment]
Changing topic - I am working at Dob:

I don't understand in what direction of a hurdle movement if the hurdle is a pawn.

In particular: https://juliasfairies.com/wp-content/uploads/ANDA-Fairy-Planet-No-1-2020.pdf, No. 10:
1. 1.dGe6-e4[bPe5->e6] # is not mentioned as solution - so the pawn movement goes according to the pawn's color (Black)
2. 1.dGe6-e8[nPe7->e6] # is not mentioned as solution - because "According to the definition, on the second step with the neutral hurdle Y, White must perceive nPe7 as a white pawn"

I haven't found anything like this in the rules, and cases 1 and 2 certainly don't seem consistent.

**Comment by vlaicu** [comment]
You are right. I have written an e-mail to Petko asking for additional clarifications and added you in Cc.
Example 10 is not consistent with the definition.

**Comment by thomas-maeder** [comment]
Back to promotions. I have looked at all the attributes:
[promotion-and-piece-attributes.out.txt](https://github.com/thomas-maeder/popeye/files/14472243/promotion-and-piece-attributes.out.txt)
[promotion-and-piece-attributes.inp.txt](https://github.com/thomas-maeder/popeye/files/14472244/promotion-and-piece-attributes.inp.txt)

Normally, a pawn with attribute A promotes into an officer with the same attribute A.

The only exception is the attribute HurdleColourChanging:
- pawns can't be HurdleColourChanging
- an "unnormal" promotion behavior is therefore required
- this behavior is: if HurdleColourChanging hoppers of some type are present, pawns can promote into HurdleColourChanging hoppers of that type, but not in regular hoppers of that type

I wasn't aware of this. I wasn't involved in the implementation of the HurdleColourChanging attribute, and I can't remember having composed a problem with such hoppers. My only exposure to them was that I directed the Quick Composing Ty in Bern 2014.

I totally agree with how Popeye deals with attributes that are applicable to all pieces.

As for Bul and Dob, it seems reasonable to treat them like HurdleColourChanging.

**Comment by dturevski** [comment]
What about Chameleon? Popeye allows chameleon pawns for some reason.

**Comment by thomas-maeder** [comment]
According to [promotion-and-piece-attributes.out.txt](https://github.com/thomas-maeder/popeye/files/14472243/promotion-and-piece-attributes.out.txt), a Chameleon pawn promotes into Chameleon officers.

**Comment by dturevski** [comment]
Yeah, I get that, the point is that it isn't consistent with how the HurdleColourChanging (and provisionally Bul/Dob) pawns are treated.

**Comment by thomas-maeder** [comment]
I see. We have less freedom here, since there may be problems whose correctness relies on the current behavior of Popeye.

**Comment by dturevski** [comment]
I respectfully disagree. Popeye at its current state cannot be used to prove correctness of a chess problem. On the other hand it is a super useful tool to prove an incorrectness. Therefore, modifying Popeye in no way can damage already published correct problems.

**Comment by thomas-maeder** [comment]
@dturevski Of course, my wording wasn't lucky. But still, changing behavior that users are used to requires more consideration than making decisions for a new feature.
